Hola **Veronica**!

Soy **Patricio Requena** 👋. Es un placer ser el revisor de tu proyecto el día de hoy!

Revisaré tu proyecto detenidamente con el objetivo de ayudarte a mejorar y perfeccionar tus habilidades. Durante mi revisión, identificaré áreas donde puedas hacer mejoras en tu código, señalando específicamente qué y cómo podrías ajustar para optimizar el rendimiento y la claridad de tu proyecto. Además, es importante para mí destacar los aspectos que has manejado excepcionalmente bien. Reconocer tus fortalezas te ayudará a entender qué técnicas y métodos están funcionando a tu favor y cómo puedes aplicarlos en futuras tareas. 

_**Recuerda que al final de este notebook encontrarás un comentario general de mi parte**_, empecemos!

Encontrarás mis comentarios dentro de cajas verdes, amarillas o rojas, ⚠️ **por favor, no muevas, modifiques o borres mis comentarios** ⚠️:


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class=“tocSkip”></a>
Si todo está perfecto.
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class=“tocSkip”></a>
Si tu código está bien pero se puede mejorar o hay algún detalle que le hace falta.
</div>

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class=“tocSkip”></a>
Si de pronto hace falta algo o existe algún problema con tu código o conclusiones.
</div>

Puedes responderme de esta forma:
<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class=“tocSkip”></a>


<div class="alert alert-block alert-warning">
<b>Comentario del revisor (1ra Iteracion)</b> <a class=“tocSkip”></a>

Para mantener buenas prácticas y una correcta presentación en tus notebooks te recomiendo siempre dejar al inicio de los mismos en una celda markdown un título descriptivo y una breve introducción al proyecto.

</div>

Sprint 11 — Predicción de Churn (Beta Bank)

En este proyecto se construye un modelo de clasificación para predecir si un cliente abandonará el banco (Exited = 1). 
El objetivo principal es maximizar el F1-score (mínimo requerido: 0.59) y complementar la evaluación con AUC-ROC.

Pasos de trabajo:

1.- Carga y preparación de datos (limpieza, tratamiento de valores faltantes y codificación de variables categóricas).

2.- Revisión del equilibrio de clases y entrenamiento de un modelo base sin corrección del desbalance.

3.- Mejora del modelo aplicando al menos dos técnicas de corrección del desbalance (pesos de clase y re-muestreo), selección del mejor modelo usando validación.

4.- Prueba final del modelo en el conjunto de prueba y reporte de métricas finales (F1 y AUC-ROC).

In [1]:
import pandas as pd
data = pd.read_csv('/datasets/Churn.csv')
data.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0     2.0       0.00              1          1               1   
1     1.0   83807.86              1          0               1   
2     8.0  159660.80              3          1               0   
3     1.0       0.00              2          0               0   
4     2.0  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

In [2]:
data.info()
data.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           9091 non-null   float64
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


RowNumber    CustomerId   CreditScore           Age       Tenure  \
count  10000.00000  1.000000e+04  10000.000000  10000.000000  9091.000000   
mean    5000.50000  1.569094e+07    650.528800     38.921800     4.997690   
std     2886.89568  7.193619e+04     96.653299     10.487806     2.894723   
min        1.00000  1.556570e+07    350.000000     18.000000     0.000000   
25%     2500.75000  1.562853e+07    584.000000     32.000000     2.000000   
50%     5000.50000  1.569074e+07    652.000000     37.000000     5.000000   
75%     7500.25000  1.575323e+07    718.000000     44.000000     7.000000   
max    10000.00000  1.581569e+07    850.000000     92.000000    10.000000   

             Balance  NumOfProducts    HasCrCard  IsActiveMember  \
count   10000.000000   10000.000000  10000.00000    10000.000000   
mean    76485.889288       1.530200      0.70550        0.515100   
std     62397.405202       0.581654      0.45584        0.499797   
min         0.000000       1.000000      0.00000        0.000000   
25%         0.000000       1.000000      0.00000        0.000000   
50%     97198.540000       1.000000      1.00000        1.000000   
75%    127644.240000       2.000000      1.00000        1.000000   
max    250898.090000       4.000000      1.00000        1.000000   

       EstimatedSalary        Exited  
count     10000.000000  10000.000000  
mean     100090.239881      0.203700  
std       57510.492818      0.402769  
min          11.580000      0.000000  
25%       51002.110000      0.000000  
50%      100193.915000      0.000000  
75%      149388.247500      0.000000  
max      199992.480000      1.000000

El conjunto de datos se cargó directamente desde la ruta proporcionada por la plataforma (/datasets/Churn.csv) utilizando pandas. 
Posteriormente, se realizó una revisión inicial del dataset para verificar dimensiones, tipos de datos y consistencia general de la información antes de iniciar el preprocesamiento.

In [3]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import f1_score, roc_auc_score, classification_report
from sklearn.utils import shuffle


In [4]:
data = pd.read_csv('/datasets/Churn.csv')
data.head()


RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0     2.0       0.00              1          1               1   
1     1.0   83807.86              1          0               1   
2     8.0  159660.80              3          1               0   
3     1.0       0.00              2          0               0   
4     2.0  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

In [5]:
data.info()
data['Exited'].value_counts(normalize=True)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           9091 non-null   float64
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


0    0.7963
1    0.2037
Name: Exited, dtype: float64

In [6]:
data = data.drop(['RowNumber', 'CustomerId', 'Surname'], axis=1)
data['Tenure'] = data['Tenure'].fillna(data['Tenure'].median())
data = pd.get_dummies(data, drop_first=True)


In [7]:
X = data.drop('Exited', axis=1)
y = data['Exited']


In [8]:
X_train, X_temp, y_train, y_temp = train_test_split(
    X, y, test_size=0.4, random_state=12345, stratify=y
)

X_valid, X_test, y_valid, y_test = train_test_split(
    X_temp, y_temp, test_size=0.5, random_state=12345, stratify=y_temp
)


In [9]:
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_valid_scaled = scaler.transform(X_valid)
X_test_scaled = scaler.transform(X_test)


In [10]:
model = LogisticRegression(random_state=12345, max_iter=1000)
model.fit(X_train_scaled, y_train)

y_pred = model.predict(X_valid_scaled)
y_proba = model.predict_proba(X_valid_scaled)[:, 1]

print("F1 (valid):", f1_score(y_valid, y_pred))
print("AUC-ROC (valid):", roc_auc_score(y_valid, y_proba))
print(classification_report(y_valid, y_pred))


F1 (valid): 0.3107861060329068
AUC-ROC (valid): 0.7874451916444971
              precision    recall  f1-score   support

           0       0.83      0.97      0.89      1592
           1       0.61      0.21      0.31       408

    accuracy                           0.81      2000
   macro avg       0.72      0.59      0.60      2000
weighted avg       0.78      0.81      0.77      2000



<div class="alert alert-block alert-danger">
<b>Comentario del revisor (1ra Iteracion)</b> <a class=“tocSkip”></a>

La ventaja de usar notebooks es que puedas separar cada proceso en celdas diferentes, tener todo un solo proceso en una celda extensa no es una buena práctica ya que dificulta la lectura y seguimiento del proceso. Por favor, cada parte de tu código debe ser separada para un correcto seguimiento.

Además, no es buena práctica el cargar los datos más de una vez ya que se pueden perder procesos aplicados al dataframe. Sólo se debe cargar una vez y a partir de ahí hacer uso de las variables.
</div>

En esta sección se evaluó el equilibrio de clases de la variable objetivo Exited, con el objetivo de identificar si existe un desbalance que pueda afectar el desempeño del modelo.

Primero se revisó la proporción de clientes que permanecen en el banco (0) versus los que abandonan (1). Se observa un desequilibrio de clases, donde la clase 0 es mayoritaria y la clase 1 es minoritaria. Este desbalance puede provocar que el modelo favorezca la clase mayoritaria.

Tras entrenar un modelo base de Regresión Logística sin aplicar técnicas de balanceo, se evaluó su desempeño en el conjunto de validación.

Los resultados muestran:

Accuracy ≈ 0.81

F1-score global ≈ 0.31

Recall para la clase 1 ≈ 0.21

Aunque la accuracy parece elevada, el bajo recall y F1 para la clase minoritaria indican que el modelo no detecta correctamente a los clientes que abandonan el banco. Esto confirma que el desequilibrio de clases afecta negativamente el desempeño y debe corregirse en la siguiente etapa.

En la etapa de preparación de datos ya se realizó la limpieza necesaria, incluyendo la imputación de valores faltantes y la eliminación de variables no predictivas, por lo que en esta sección se trabaja directamente con los datos preparados.

<div class="alert alert-block alert-success">
<b>Comentario del revisor (2da Iteracion)</b> <a class=“tocSkip”></a>

Muy bien! Siempre es importante seguir buenas prácticas en el uso de notebooks ya que en el entorno laboral a menudo trabajarás con más personas y para que se de un correcto seguimiento y entendimiento de tu trabajo es importante que sea claro y fácil de seguir.
</div>

In [11]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import f1_score, roc_auc_score

results = []

def evaluate_model(model, X_tr, y_tr, X_val, y_val, model_name="model"):
    model.fit(X_tr, y_tr)
    preds = model.predict(X_val)
    proba = model.predict_proba(X_val)[:, 1]
    f1 = f1_score(y_val, preds)
    auc = roc_auc_score(y_val, proba)
    return {"model": model_name, "F1_valid": f1, "AUC_valid": auc, "estimator": model}

for C in [0.01, 0.1, 1, 10]:
    lr_bal = LogisticRegression(
        random_state=12345,
        max_iter=2000,
        class_weight='balanced',
        C=C,
        solver='lbfgs'
    )
    res = evaluate_model(lr_bal, X_train_scaled, y_train, X_valid_scaled, y_valid,
                         model_name=f"LR_balanced_C={C}")
    results.append(res)
    print(res["model"], "| F1:", round(res["F1_valid"], 4), "| AUC:", round(res["AUC_valid"], 4))

print("\nTotal resultados guardados:", len(results))


LR_balanced_C=0.01 | F1: 0.5295 | AUC: 0.7936
LR_balanced_C=0.1 | F1: 0.5285 | AUC: 0.7936
LR_balanced_C=1 | F1: 0.5281 | AUC: 0.7937
LR_balanced_C=10 | F1: 0.5281 | AUC: 0.7937

Total resultados guardados: 4


Con class_weight='balanced', el mejor F1 en validación fue ~0.53 (AUC ~0.79). Variar C no cambió materialmente el desempeño.

In [12]:
# ==============================
# Función de evaluación común
# ==============================

def evaluate_model(model, X_tr, y_tr, X_val, y_val, model_name="model"):
    model.fit(X_tr, y_tr)
    preds = model.predict(X_val)
    proba = model.predict_proba(X_val)[:, 1]
    
    f1 = f1_score(y_val, preds)
    auc = roc_auc_score(y_val, proba)
    
    return {
        "model": model_name,
        "F1_valid": f1,
        "AUC_valid": auc,
        "estimator": model
    }


In [13]:
results = []


In [14]:
# =========================
# ENFOQUE 1: class_weight='balanced'
# =========================

# 1A) Logistic Regression con class_weight
# Buscamos mejor C en validación (C controla regularización)
for C in [0.01, 0.1, 1, 10]:
    lr_bal = LogisticRegression(
        random_state=12345,
        max_iter=1000,
        class_weight='balanced',
        C=C,
        solver='lbfgs'
    )
    res = evaluate_model(lr_bal, X_train_scaled, y_train, X_valid_scaled, y_valid,
                         model_name=f"LR_balanced_C={C}")
    results.append(res)

# 1B) RandomForest con class_weight
# Nota: RandomForest no requiere escalado, pero podemos usar las versiones escaladas sin problema
# (no afecta, solo es redundante). Si quieres, puedes usar X_train y X_valid sin escalar.
for depth in [3, 5, 7, None]:
    for n_estimators in [100, 200]:
        rf_bal = RandomForestClassifier(
            random_state=12345,
            n_estimators=n_estimators,
            max_depth=depth,
            class_weight='balanced',
            n_jobs=-1
        )
        res = evaluate_model(rf_bal, X_train_scaled, y_train, X_valid_scaled, y_valid,
                             model_name=f"RF_balanced_depth={depth}_n={n_estimators}")
        results.append(res)



<div class="alert alert-block alert-danger">
<b>Comentario del revisor (1ra Iteracion)</b> <a class=“tocSkip”></a>

Esta celda no se puede ejecutar ya que la clase `RandomForestClassifier` no ha sido importada lo que causa un error de que no ha sido definido.

Siempre para identificar estos errores reinicia el kernel de tu notebook y ejecuta todo desde cero.
</div>

In [15]:
# Ver cuántos resultados se guardaron
print("Total resultados guardados:", len(results))

# Mostrar top por F1
results_df = pd.DataFrame([{
    "model": r["model"],
    "F1_valid": r["F1_valid"],
    "AUC_valid": r["AUC_valid"]
} for r in results]).sort_values("F1_valid", ascending=False)

print(results_df.head(10))


Total resultados guardados: 12
                           model  F1_valid  AUC_valid
9      RF_balanced_depth=7_n=200  0.641711   0.870545
8      RF_balanced_depth=7_n=100  0.632911   0.870646
7      RF_balanced_depth=5_n=200  0.622568   0.864737
6      RF_balanced_depth=5_n=100  0.618861   0.865238
5      RF_balanced_depth=3_n=200  0.596878   0.854407
4      RF_balanced_depth=3_n=100  0.588768   0.852290
11  RF_balanced_depth=None_n=200  0.587879   0.863964
10  RF_balanced_depth=None_n=100  0.578387   0.862270
0             LR_balanced_C=0.01  0.529464   0.793551
1              LR_balanced_C=0.1  0.528534   0.793650


Enfoque 1: Corrección del desequilibrio mediante class_weight='balanced'

En este primer enfoque se corrigió el desequilibrio de clases ajustando los pesos de la clase minoritaria mediante el parámetro class_weight='balanced'. Se entrenaron distintos modelos y combinaciones de hiperparámetros utilizando los conjuntos de entrenamiento y validación.

Los resultados muestran que la Regresión Logística, aun con pesos balanceados, alcanza un F1 cercano a 0.53, lo cual representa una mejora respecto al modelo base, pero no resulta suficiente para cumplir el umbral requerido del proyecto.

En contraste, los modelos Random Forest con pesos balanceados presentan un desempeño significativamente superior. El mejor modelo obtenido en este enfoque alcanzó un F1 ≈ 0.64 y un AUC-ROC ≈ 0.87 en el conjunto de validación, lo que indica una mejora clara en la detección de clientes que abandonan el banco.

Aunque este enfoque ya ofrece un desempeño sólido, se decidió explorar un segundo método de corrección del desequilibrio para comparar resultados y evaluar la estabilidad del modelo.

In [16]:
# =========================
# ENFOQUE 2: Re-muestreo (Upsampling) del train
# =========================

def upsample_train(X_tr, y_tr, random_state=12345):
    """
    Upsampling: duplicamos la clase minoritaria hasta igualar la mayoritaria.
    OJO: se aplica SOLO al entrenamiento para evitar data leakage.
    """
    X_tr_df = pd.DataFrame(X_tr)  # por si X_tr viene como numpy array (es tu caso: scaled)
    y_tr_s = pd.Series(y_tr).reset_index(drop=True)
    X_tr_df = X_tr_df.reset_index(drop=True)

    # separar clases
    X_0 = X_tr_df[y_tr_s == 0]
    X_1 = X_tr_df[y_tr_s == 1]

    # identificar mayoritaria/minoritaria
    if len(X_1) < len(X_0):
        X_min, y_min = X_1, np.ones(len(X_1), dtype=int)
        X_maj, y_maj = X_0, np.zeros(len(X_0), dtype=int)
    else:
        X_min, y_min = X_0, np.zeros(len(X_0), dtype=int)
        X_maj, y_maj = X_1, np.ones(len(X_1), dtype=int)

    # re-muestreo con reemplazo para igualar tamaños
    X_min_upsampled = X_min.sample(n=len(X_maj), replace=True, random_state=random_state)

    # reconstruir dataset balanceado
    X_bal = pd.concat([X_maj, X_min_upsampled], axis=0)
    y_bal = np.concatenate([y_maj, np.full(len(X_min_upsampled), 1 if (len(X_1) < len(X_0)) else 0)])

    # barajar
    X_bal, y_bal = shuffle(X_bal, y_bal, random_state=random_state)

    return X_bal.values, y_bal


<div class="alert alert-block alert-success">
<b>Comentario del revisor (1ra Iteracion)</b> <a class=“tocSkip”></a>

Perfecto, técnicas cómo sobre-muestreo o sub-muestreo o la modificación de parámetros en algunos modelos para especificar el desbalance ayuda a los modelos a poder generalizar mejor cada caso y obtener un buen desempeño
</div>

In [17]:
X_train_up, y_train_up = upsample_train(X_train_scaled, y_train)

print("Distribución original (train):")
print(pd.Series(y_train).value_counts())

print("\nDistribución con upsampling (train):")
print(pd.Series(y_train_up).value_counts())

for C in [0.01, 0.1, 1, 10]:
    lr_up = LogisticRegression(random_state=12345, max_iter=2000, C=C, solver='lbfgs')
    lr_up.fit(X_train_up, y_train_up)

    preds = lr_up.predict(X_valid_scaled)
    proba = lr_up.predict_proba(X_valid_scaled)[:, 1]

    f1 = f1_score(y_valid, preds)
    auc = roc_auc_score(y_valid, proba)

    print(f"LR_upsampling_C={C} | F1: {f1:.4f} | AUC: {auc:.4f}")

best_up = {"name": None, "f1": 0, "auc": 0, "model": None}

for depth in [3, 5, 7, None]:
    for n in [100, 200]:
        rf_up = RandomForestClassifier(
            random_state=12345,
            n_estimators=n,
            max_depth=depth,
            n_jobs=-1
        )
        rf_up.fit(X_train_up, y_train_up)

        preds = rf_up.predict(X_valid_scaled)
        proba = rf_up.predict_proba(X_valid_scaled)[:, 1]

        f1 = f1_score(y_valid, preds)
        auc = roc_auc_score(y_valid, proba)

        name = f"RF_upsampling_depth={depth}_n={n}"
        print(f"{name} | F1: {f1:.4f} | AUC: {auc:.4f}")

        if f1 > best_up["f1"]:
            best_up = {"name": name, "f1": f1, "auc": auc, "model": rf_up}

print("\nMEJOR MODELO UPSAMPLING (valid):")
print(best_up["name"], "| F1:", round(best_up["f1"], 4), "| AUC:", round(best_up["auc"], 4))


Distribución original (train):
0    4778
1    1222
Name: Exited, dtype: int64

Distribución con upsampling (train):
0    4778
1    4778
dtype: int64
LR_upsampling_C=0.01 | F1: 0.5217 | AUC: 0.7908
LR_upsampling_C=0.1 | F1: 0.5213 | AUC: 0.7909
LR_upsampling_C=1 | F1: 0.5213 | AUC: 0.7910
LR_upsampling_C=10 | F1: 0.5213 | AUC: 0.7910
RF_upsampling_depth=3_n=100 | F1: 0.5939 | AUC: 0.8538
RF_upsampling_depth=3_n=200 | F1: 0.5953 | AUC: 0.8530
RF_upsampling_depth=5_n=100 | F1: 0.6191 | AUC: 0.8667
RF_upsampling_depth=5_n=200 | F1: 0.6157 | AUC: 0.8667
RF_upsampling_depth=7_n=100 | F1: 0.6366 | AUC: 0.8721
RF_upsampling_depth=7_n=200 | F1: 0.6420 | AUC: 0.8725
RF_upsampling_depth=None_n=100 | F1: 0.6060 | AUC: 0.8600
RF_upsampling_depth=None_n=200 | F1: 0.6095 | AUC: 0.8628

MEJOR MODELO UPSAMPLING (valid):
RF_upsampling_depth=7_n=200 | F1: 0.642 | AUC: 0.8725


<div class="alert alert-block alert-danger">
<b>Comentario del revisor (1ra Iteracion)</b> <a class=“tocSkip”></a>

`NumPy` no ha sido importando, causando un error en la ejecución de esta celda en específico.
</div>

<div class="alert alert-block alert-success">
<b>Comentario del revisor (2da Iteracion)</b> <a class=“tocSkip”></a>

Correcto, en proyectos de clasificación siempre es importante revisar el equilibrio de las clases de nuestra variable objetivo y así de esta forma determinar la mejor estrategia o modelo para mitigar el impacto del desbalance que se pueda tener.
</div>

Enfoque 2: Corrección del desequilibrio mediante re-muestreo (upsampling)

En este segundo enfoque se corrigió el desequilibrio de clases mediante upsampling del conjunto de entrenamiento, duplicando la clase minoritaria hasta igualar el tamaño de la clase mayoritaria. Este procedimiento se aplicó exclusivamente al conjunto de entrenamiento para evitar data leakage.

Tras el balanceo, se entrenaron nuevamente modelos de Regresión Logística y Random Forest, evaluando su desempeño en el conjunto de validación.

Los resultados muestran que la Regresión Logística no presenta mejoras relevantes en F1 al aplicar upsampling. En cambio, los modelos Random Forest entrenados con el conjunto balanceado alcanzan un desempeño superior.

El mejor modelo de este enfoque obtuvo un F1 ≈ 0.64 y un AUC-ROC ≈ 0.87, valores comparables a los obtenidos en el Enfoque 1 con pesos balanceados, lo que indica consistencia en el desempeño del modelo.


In [18]:
# =========================
# COMPARACIÓN Y SELECCIÓN DEL MEJOR (por F1 en validación)
# =========================

results_df = pd.DataFrame([{"model": r["model"], "F1_valid": r["F1_valid"], "AUC_valid": r["AUC_valid"]} for r in results])
results_df = results_df.sort_values(by="F1_valid", ascending=False)

display(results_df.head(10))

# Tomamos el mejor modelo por F1 en validación
best_row = results_df.iloc[0]
best_name = best_row["model"]

best_model = next(r["estimator"] for r in results if r["model"] == best_name)

print("Mejor modelo (valid):", best_name)
print("F1_valid:", best_row["F1_valid"])
print("AUC_valid:", best_row["AUC_valid"])



model  F1_valid  AUC_valid
9      RF_balanced_depth=7_n=200  0.641711   0.870545
8      RF_balanced_depth=7_n=100  0.632911   0.870646
7      RF_balanced_depth=5_n=200  0.622568   0.864737
6      RF_balanced_depth=5_n=100  0.618861   0.865238
5      RF_balanced_depth=3_n=200  0.596878   0.854407
4      RF_balanced_depth=3_n=100  0.588768   0.852290
11  RF_balanced_depth=None_n=200  0.587879   0.863964
10  RF_balanced_depth=None_n=100  0.578387   0.862270
0             LR_balanced_C=0.01  0.529464   0.793551
1              LR_balanced_C=0.1  0.528534   0.793650

Mejor modelo (valid): RF_balanced_depth=7_n=200
F1_valid: 0.6417112299465241
AUC_valid: 0.870544511774559


Importantes hallazgos:

class_weight='balanced' normalmente sube el recall/F1 de la clase 1 sin tocar la data; es una corrección “limpia”.

Upsampling suele subir F1 también, pero puede incrementar falsos positivos.

RandomForest a veces supera a la logística, pero depende del tuning; la logística suele ser una base fuerte y fácil de explicar.


Resumen de ambos enfoques y su comparación
En esta etapa se evaluaron distintos enfoques para mejorar la calidad del modelo, abordando explícitamente el desequilibrio de clases presente en la variable objetivo Exited. Se trabajó con conjuntos separados de entrenamiento y validación para seleccionar el mejor modelo y conjunto de hiperparámetros antes de realizar la prueba final.

Enfoque 1: class_weight='balanced'

En el primer enfoque se corrigió el desequilibrio ajustando los pesos de la clase minoritaria mediante el parámetro class_weight='balanced'.
Se entrenaron modelos de Regresión Logística y Random Forest, probando distintas combinaciones de hiperparámetros sobre los conjuntos de entrenamiento y validación.

Los resultados muestran que la Regresión Logística, aun con pesos balanceados, alcanza un F1 cercano a 0.53, lo cual representa una mejora respecto al modelo base, pero no es suficiente para cumplir el umbral requerido por el proyecto.

En contraste, los modelos Random Forest con pesos balanceados presentan un desempeño significativamente superior. El mejor modelo de este enfoque alcanzó un F1 ≈ 0.64 y un AUC-ROC ≈ 0.87 en el conjunto de validación, lo que indica una mejor capacidad para detectar clientes que abandonan el banco.

Enfoque 2: Re-muestreo del conjunto de entrenamiento (Upsampling)

En el segundo enfoque se corrigió el desequilibrio mediante upsampling del conjunto de entrenamiento, duplicando la clase minoritaria hasta igualar el tamaño de la clase mayoritaria.
Este procedimiento se aplicó exclusivamente al conjunto de entrenamiento para evitar data leakage.

Tras el balanceo, se entrenaron nuevamente modelos de Regresión Logística y Random Forest. Los resultados muestran que la Regresión Logística no presenta mejoras relevantes en F1 al aplicar upsampling. Sin embargo, los modelos Random Forest entrenados con el conjunto balanceado alcanzan un desempeño sólido, con un mejor F1 ≈ 0.64 y AUC-ROC ≈ 0.87 en validación.

Estos valores son comparables a los obtenidos en el enfoque con pesos balanceados, lo que indica consistencia en el rendimiento del modelo.

Comparación y selección del mejor modelo

Al comparar todos los modelos evaluados utilizando el F1-score como métrica principal, el mejor desempeño en validación corresponde a un modelo Random Forest con class_weight='balanced' y profundidad máxima igual a 7, con 200 estimadores.

Este modelo logra el mayor equilibrio entre detección de la clase minoritaria y estabilidad general, por lo que se selecciona como el modelo final para su evaluación en el conjunto de prueba.

In [19]:
# ==========================
# PRUEBA FINAL EN TEST
# ==========================

# Predicciones en test
y_test_pred = best_model.predict(X_test_scaled)
y_test_proba = best_model.predict_proba(X_test_scaled)[:, 1]

# Métricas finales
f1_test = f1_score(y_test, y_test_pred)
auc_test = roc_auc_score(y_test, y_test_proba)

print("Resultados finales en TEST:")
print("F1 (test):", round(f1_test, 4))
print("AUC-ROC (test):", round(auc_test, 4))


Resultados finales en TEST:
F1 (test): 0.606
AUC-ROC (test): 0.8555


Prueba final del modelo

En la etapa final se evaluó el modelo seleccionado utilizando el conjunto de prueba, el cual no fue utilizado durante el entrenamiento ni la validación. Esto permite estimar de forma realista el desempeño del modelo frente a datos no vistos.

El modelo evaluado fue un Random Forest con class_weight='balanced', profundidad máxima de 7 y 200 estimadores, seleccionado por obtener el mejor F1-score en el conjunto de validación.

Los resultados obtenidos en el conjunto de prueba fueron:

F1-score (test): 0.606

AUC-ROC (test): 0.8555

El valor de F1 supera el umbral mínimo requerido de 0.59, lo que indica que el modelo identifica adecuadamente a los clientes que abandonan el banco, equilibrando precisión y recall en la clase minoritaria.

Por su parte, el AUC-ROC confirma una buena capacidad discriminativa del modelo para separar clientes que permanecen de aquellos que abandonan, y es consistente con los resultados observados durante la validación.

Conclusión general

A lo largo del proyecto se prepararon y limpiaron los datos, se analizó el desequilibrio de clases y se entrenaron distintos modelos utilizando múltiples estrategias de corrección. Se compararon los modelos mediante métricas apropiadas para problemas desbalanceados, priorizando el F1-score.

El modelo final cumple con los criterios establecidos y generaliza correctamente a datos no vistos, por lo que se considera adecuado para el objetivo de predicción de abandono de clientes.

Recomendaciones

El modelo muestra un desempeño adecuado para identificar clientes con alta probabilidad de abandono. Dado que el costo de perder un cliente suele ser mayor que el costo de intervenir preventivamente, se recomienda priorizar el recall de la clase minoritaria, incluso a costa de un ligero aumento en falsos positivos.

Para un uso operativo, el modelo podría complementarse con un ajuste del umbral de decisión, permitiendo adaptar la sensibilidad del sistema según la estrategia del banco (retención agresiva vs. conservadora).

En un escenario productivo, sería recomendable reentrenar periódicamente el modelo y monitorear el desempeño de F1 y AUC-ROC para detectar posibles cambios en el comportamiento de los clientes.

<div class="alert alert-block alert-danger">
<b>Comentario del revisor (1ra Iteracion)</b> <a class=“tocSkip”></a>

Hay varios módulos que no fueron importados lo que impide la ejecución total del notebook, por favor revisar y ejecutar todo desde cero reiniciando el kernel para realizar esas correcciones.

Saludos!
</div>

<div class="alert alert-block alert-success">
<b>Comentario del revisor (2da Iteracion)</b> <a class=“tocSkip”></a>

Te felicito por el trabajo realizado Veronica, se nota que conoces las métricas de evaluación y planteas conclusiones muy acertadas en cuanto a los resultados. Es importante siempre en un problema de ML identificar el balance entre las clases de tu variable objetivo para poder escoger la métrica correcta ya que para datasets desbalanceados la métrica de accuracy suele presentar buenos valores pero en realidad las predicciones no son buenas y es por la forma en cómo se calcula esta métrica que puede llevar a interpretaciones erroneas en cambio el f1-score es la ideal para este tipo de casos ya que maneja mejor este tipo de problemas.
    
Saludos!
</div>